os/drivers/input/touch.c

```diff
@@ -0,0 +1,325 @@
+ ************************************************************************************/
+
+static inline int touch_semtake(FAR sem_t *sem, bool errout)
+{
+	/* Take a count from the semaphore, possibly waiting */
+	if (sem_wait(sem) < 0) {
+		/* EINTR is the only error that we expect */
+		int errcode = get_errno();
+		DEBUGASSERT(errcode == EINTR);
+		if (errout) {
+			return errcode;
+		}
+	}
+	return OK;
+}
+
+/****************************************************************************
+ * Name: touch_semgive
+ ****************************************************************************/
+
+static inline void touch_semgive(sem_t *sem)
+{
+	sem_post(sem);
+}
+
+static int touch_open(FAR struct file *filep)
+{
+	FAR struct touch_s *priv;
+	priv = filep->f_inode->i_private;
+
+	if (!priv) {
+		return -EINVAL;
+	}
+
+	touch_semtake(&priv->sem, false);
+	if (priv->crefs == 0) {
+		priv->dev->touch_enable(priv->dev, true);
+	}
+	priv->crefs++;
+	DEBUGASSERT(priv->crefs > 0);
+	touch_semgive(&priv->sem);
+	return OK;
+}
+
+static int touch_close(FAR struct file *filep)
+{
+	FAR struct touch_s *priv;
+	priv = filep->f_inode->i_private;
+
+	if (!priv) {
+		return -EINVAL;
+	}
+
+	touch_semtake(&priv->sem, false);
+	DEBUGASSERT(priv->crefs > 0);
+	priv->crefs--;
+	if (priv->crefs == 0) {
+		priv->dev->touch_enable(priv->dev, false);
+#ifndef CONFIG_DISABLE_POLL
+		/* Check if this file is registered in a list of waiters for polling.
+		* For example, when task A is blocked by calling poll and task B try to terminate task A,
+		* a pollfd of A remains in this list. If it is, it should be cleared.
+		*/
+		(void)touch_semtake(&priv->pollsem, false);
+		for (int i = 0; i < CONFIG_TOUCH_NPOLLWAITERS; i++) {
+			struct pollfd *fds = priv->fds[i];
+			if (fds && (FAR struct file *)fds->filep == filep) {
+				priv->fds[i] = NULL;
+			}
+		}
+		touch_semgive(&priv->pollsem);
+#endif
+	}
+	touch_semgive(&priv->sem);
+	return OK;
+}
+
+static ssize_t touch_read(FAR struct file *filep, FAR char *buffer, size_t buflen)
+{
+	FAR struct inode *inode;
+	FAR struct touch_s *priv;
+	size_t outlen;
+	int ret;
+
+	DEBUGASSERT(buffer);
+	DEBUGASSERT(buflen > 0);
+	DEBUGASSERT(filep);
+	inode = filep->f_inode;
+
+	DEBUGASSERT(inode && inode->i_private);
+	priv = inode->i_private;
+
+	/* Wait for semaphore to prevent concurrent reads */
+	touch_semtake(&priv->sem, false);
+
+	/* Read the touch data, only if screen has been touched or if we're waiting for touch up */
+	outlen = sizeof(struct touch_sample_s);
+	ret = priv->dev->touch_read(priv->dev, buffer);
+	touch_semgive(&priv->sem);
+	return ret < 0 ? ret : outlen;
+}
+
+static ssize_t touch_write(FAR struct file *filep, FAR const char *buffer, size_t buflen)
+{
+	//TODO
+	return OK;
+}
+
+#ifndef CONFIG_DISABLE_POLL
+void touch_pollnotify(FAR struct touch_dev_s *priv, pollevent_t eventset)
+{
+	int itr;
+	for (itr = 0; itr < CONFIG_TOUCH_NPOLLWAITERS; itr++) {
+		struct pollfd *fds = g_dev->fds[itr];
```

Comment:
you can access priv upper layer as below, let's remove global variable

```
int touch_register(struct touch_dev_s *dev) {
     dev->upper= upper;
}
```

Reply:
@namanjain7 @ewoodev How is this?